bokeh/models/annotations.py

```diff
@@ -35,19 +44,64 @@ class Annotation(Renderer):
 
     level = Override(default="annotation")
 
+
 @abstract
 class TextAnnotation(Annotation):
     """ Base class for annotation models.
 
     """
 
+
+class LegendItem(Model):
+
+    def __init__(self, *args, **kwargs):
+        super(LegendItem, self).__init__(*args, **kwargs)
+        if isinstance(self.label, string_types):
+            # Allow convenience of setting label as a string
+            self.label = value(self.label)
+
+    label = StringSpec(default=None, help="""
+    A label for this legend. Can be a string, or a column of a
+    ColumnDataSource. If ``label`` is a field, then it must
+    be in the renderers' data_source.
+    """)
+
+    renderers = List(Instance(GlyphRenderer), help="""
+    A list of the glyph renderers to draw in the legend. If ``label`` is a field,
+    then all data_sources of renderers must be the same.
+    """)
+
+    @validation.error(NON_MATCHING_DATA_SOURCES_ON_LEGEND_ITEM_RENDERERS)
+    def _check_data_sources_on_renderers(self):
+        if self.label and 'field' in self.label:
+            if len({r.data_source for r in self.renderers}) != 1:
+                return str(self)
+
+    @validation.error(BAD_COLUMN_NAME)
+    def _check_field_label_on_data_source(self):
+        if self.label and 'field' in self.label:
+            if len(self.renderers) < 1:
+                return str(self)
+            source = self.renderers[0].data_source
+            if self.label.get('field') not in source.column_names:
```

Comment:
It's not a problem with this PR, but whenever I see `prop.get('field')`, knowing that `prop` is `StringSpec` or similar, my head boils. We need to get rid of all those dictionaries that attempt to act like structs.